### DISCUSSION TITLE
how to set action env variables within a shell script?

### DISCUSSION THREAD

**Discussion by andy-brainome** [community/community#27127] [Actions]
<p>I’ve got a bash script that forms a bunch of env variables used to tag build artifacts.</p>
<p>How do I export these variables from the shell script into the action build variables?</p>
<p>(EDITED)</p>
<p>actually running </p>
<p>echo “::set-env name=VERSION_TAG::$VERSION_TAG”</p>
<p>within the shell script does does translate into an action variable.</p>
<p>I’ve built a test bed that does work :slight_smile:</p>
<p><a href="https://github.com/astevko/action-test" rel="nofollow noopener">https://github.com/astevko/action-test</a></p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/andy-brainome">@andy-brainome</a> , </p>
<p>The <strong>set-env</strong> command will add environment variable to all the following steps in the current job. </p>
<p><img src="https://user-images.githubusercontent.com/24806478/181101970-d82efb3b-e7ed-4b80-98d2-050ce28deb3f.png" alt="action env.png" data-base62-sha1="2nb4ZsmAugF9L0OjWIBFq5tEm8H" width="404" height="379"></p>
<p>What’s your meaning of “action build variables” ? If you want to use the env in action’s <strong>input</strong> under <strong>with</strong> section, you could use </p>
<p>${{ env.VERSION_TAG }} to get the value of it. For example: </p>
<pre><code>- name: Create Release
      id: create_release
      uses: actions/create-release@latest
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # This token is provided by Actions, you do not need to create your own token
      with:
        tag_name: ${{ env.VERSION_TAG }}</code></pre>

**Comment by andy-brainome** [comment]
<p>My issue is with exporting env variables from within a shell script into action env variables.</p>
<p>I tried using the echo “::set-env…”  from within the shell script without success - I suspect ::set-env does not trigger the action from registering env variables</p>
<p>pseudocode e.g.</p>
<p>calc-version.sh</p>
<pre><code>#!/bin/bash
export VERSION=19.2.5
export FULL_TAG=19.2.5-beta
export SHORT_TAG=beta
</code></pre>
<p>build-docker.yml</p>
<pre><code>&lt;snip&gt;
- name: calc version
   run: ./calc-version.sh 

- name: build and push myimage
   run: |
     docker build -t myimage:${{ env. FULL_TAG }} .
     docker push myimage:${{ env.FULL_TAG }}
     docker tag myimage:${{ env. FULL_TAG }} myimage:${{ env.SHORT_TAG }}
     docker push myimage:${{ env.SHORT_TAG }}
</code></pre>

**Comment by mixail-novikov** [comment]
<p>You should access your env vars like this</p>
<pre><code>docker build -t myimage:${FULL_TAG} .

docker build -t myimage:$FULL_TAG .
</code></pre>
<p>I guess that <code>run</code> command executes as is. So double curly braces isn’t a valid syntax for shell command, at least to access env variable. The two variants above are valid.</p>
<p>I think that the following syntax <code>${{ env. FULL_TAG }}</code> it’s a way to access variables in the YAML context. For example, pass arguments to actions.</p>

**Comment by Yanjingzhu** [comment]
<p>I tried add set-env command in bash scripts calc-version.sh : </p>
<pre><code>#!/bin/bash
echo "::set-env name=VERSION::19.2.5"
</code></pre>
<p>My workflow is : </p>
<pre><code>jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: set output
      id: set
      run: |

       chmod +x calc-version.sh
       ./calc-version.sh

    - name: use output
      run: |
        echo $VERSION
        echo ${{env.VERSION }}
</code></pre>
<p>Both invoke env variable syntax could work: </p>
<p><span class="math">VERSION&amp;nbsp; and&amp;nbsp;</span>{{env.VERSION }}</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/efc553f16a65a0cf9051c42180b8e4165fceaf17.png" data-download-href="/uploads/short-url/yd6T5sQqccZ7pWnZ9A66ge4Uq3R.png?dl=1" title="echo var.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181101972-52e44596-3574-41b0-8344-22597cbade68.png" alt="echo var.png" data-base62-sha1="yd6T5sQqccZ7pWnZ9A66ge4Uq3R" width="690" height="434" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/efc553f16a65a0cf9051c42180b8e4165fceaf17_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">echo var.png</span><span class="informations">763×480 10.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Can you try again? </p>

**Comment by HHHartmann** [comment]
<p>Thanks for documenting this. It pointed me in the right direction.<br>
Github has changed the mechanism since then though. Read here how it now works:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#environment-files" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#environment-files" target="_blank" rel="noopener nofollow ugc">Workflow commands for GitHub Actions - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#environment-files" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/actions/reference/workflow-commands-for-github-actions</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
